src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsService.java

```diff
@@ -0,0 +1,209 @@
+            .collect(groupingBy(Envelope::getContainer))
+            .forEach(this::processByContainer);
+    }
+
+    private void processByContainer(String containerName, List<Envelope> envelopes) {
+        log.info("Processing envelopes in {} container. Envelopes found: {}", containerName, envelopes.size());
+
+        try {
+            CloudBlobContainer blobContainer = blobManager.getContainer(containerName);
+
+            envelopes.forEach(envelope -> {
+                Optional<ZipFileProcessingResult> resultOption = getBlobInputStream(
+                    blobContainer,
+                    envelope.getZipFileName()
+                ).flatMap(inputStream ->
+                    processInputStream(inputStream, containerName, envelope.getZipFileName())
+                );
+
+                boolean isUploaded = resultOption
+                    .map(result -> uploadParsedZipFileName(envelope, result.getPdfs()))
+                    .orElse(false);
+
+                // only update envelope state if no errors occurred during zip file processing
+                if (resultOption.isPresent()) {
+                    envelopeProcessor.handleEvent(envelope, isUploaded ? DOC_UPLOADED : DOC_UPLOAD_FAILURE);
+                }
+            });
+        } catch (URISyntaxException | StorageException exception) {
+            log.error("Unable to get client for {} container", containerName, exception);
+        }
+    }
+
+    private Optional<BlobInputStream> getBlobInputStream(CloudBlobContainer blobContainer, String zipFileName) {
+        try {
+            return Optional.of(
+                blobContainer
+                    .getBlockBlobReference(zipFileName)
+                    .openInputStream()
+            );
+        } catch (URISyntaxException | StorageException exception) {
+            log.error(
+                "Unable to get blob input stream. Container: {}, blob: {}",
+                blobContainer.getName(),
+                zipFileName
+            );
+
+            return Optional.empty();
+        }
+    }
+
+    private Optional<ZipFileProcessingResult> processInputStream(
+        BlobInputStream blobInputStream,
+        String containerName,
+        String zipFileName
+    ) {
+        try (ZipInputStream zis = new ZipInputStream(blobInputStream)) {
+            return Optional.of(
+                zipFileProcessor.process(zis, containerName, zipFileName)
+            );
+        } catch (DocSignatureFailureException exception) {
+            log.warn(
+                "Rejecting blob - invalid signature. File: {}, Container: {}",
+                zipFileName,
+                containerName,
+                exception
+            );
+
+            handleEventRelatedError(DOC_SIGNATURE_FAILURE, containerName, zipFileName, exception.getMessage());
+
+            blobManager.tryMoveFileToRejectedContainer(zipFileName, containerName, null);
```

Comment:
is this needed here? The javadoc for the class says it's only uploading the files, not doing any validation

Reply:
since process still involves extraction etc - need to cover this extra bit even though it is most likely not going to be breached. but let's be wild and assume blob can be tampered during the window of creating envelope and uploading pdfs :shrug: 

note. once signature feature will be removed - this section will be gone as well. so, need to keep it for the time being

Reply:
We shouldn't do any rejections in this job. At this stage of the process envelopes have already been verified and if they've been modified since then, that shouldn't lead to a rejection, but an error, which we should be aware of, so that we could decide what to do with the envelope. It shouldn't be covered by the automated rejection process. We can prevent tampering with envelopes by restricting permissions in SAS tokens (they will prevent much more bad things). I think in this case DocSignatureFailureException should be treated just like any exception.